apps/meteor/server/services/room/service.ts

```diff
@@ -106,14 +106,19 @@ export class RoomService extends ServiceClassInternal implements IRoomService {
 			throw new MeteorError('error-not-allowed', 'Not allowed', { method: 'joinRoom' });
 		}
 
-		if (
-			isRoomWithJoinCode(room) &&
-			(!joinCode || joinCode !== room.joinCode) &&
-			!(await Authorization.hasPermission(user._id, 'join-without-join-code'))
-		) {
-			throw new MeteorError('error-code-invalid', 'Invalid Room Password', {
-				method: 'joinRoom',
-			});
+		if (isRoomWithJoinCode(room)) {
+			let fetchedJoinCode = room.joinCode;
+			if (!room.joinCode) {
+				fetchedJoinCode = (
+					(await Rooms.findOneById(room._id, {
+						projection: { joinCode: 1 },
+					})) as IRoomWithJoinCode
+				).joinCode;
+			}
+
+			if ((!joinCode || joinCode !== fetchedJoinCode) && !(await Authorization.hasPermission(user._id, 'join-without-join-code'))) {
+				throw new MeteorError('error-code-required', 'Code required', { method: 'joinRoom' });
```

Comment:
Instead of loading the joinCode to memory, just do a query by room id and join code to see if the record exists.